# Desafío Técnico - Mercado Libre

#### Ezequiel Panzarasa

# Análisis exploratorio de datos

## Armado de Dataset
Después de explorar un poco la API pública de Mercado Libre, se decidió optar por el armado de un dataset de 10.000 registros. Siendo que la API tiene un offset máximo de 1000, se decidió tomar las 10 categorías más populares (de acuerdo a la página principal de MercadoLibre. Ver imagen adjunta) y armar el dataset con 1.000 registros de cada categoría. <br>
Se eligieron estas 10 categorías siguiendo la hipótesis de que a mayor popularidad se tiene un mayor número de ventas, y más representativa resulta la información obtenida en la muestra para cada una de las publicaciones. 

In [18]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", 103)

##### Ver armado de dataset en MLC_03_RAW_DATASET.ipynb

In [19]:
raw_dataset_df = pd.read_csv("raw_dataset.csv", header=0, index_col=0, sep='|')
raw_dataset_df

/home/eze/Desktop/Panza/virtual_envs/UWML_spec/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3145: DtypeWarning: Columns (102) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


id site_id  \
0      MLA900810750     MLA   
1      MLA856313404     MLA   
2      MLA618887655     MLA   
3      MLA855055612     MLA   
4      MLA841426834     MLA   
...             ...     ...   
10478  MLA678679714     MLA   
10479  MLA858821035     MLA   
10480  MLA861128302     MLA   
10481  MLA818759510     MLA   
10482  MLA857013842     MLA   

                                                   title     price  \
0          Repetidor Tp-link Tl-wa850re  Blanco 1 Unidad   2055.00   
1      Disco Sólido Interno Kingston Sa400s37/240g 240gb   3688.00   
2      Pc Armada Cpu Amd Full Gamer A10-9700 16gb Ddr...  53999.00   
3      Notebook Exo Smart E19 Intel Celeron Ram4gb Ss...  39999.00   
4      Pasta Térmica Arctic Mx-4 4grs. Alto Rendimien...   1600.00   
...                                                  ...       ...   
10478  Auriculares Genius Hs 05a Headset Control Volu...   2249.00   
10479  The Lego Movie Videogame Formato Físico Ps4 Or...   2999.00   
10480  Call Of Duty Black Ops 2 Ps3 Gold | Español | ...    449.95   
10481  Cable Cargador Y Transferencia Ps Vita Alto Re...    388.99   
10482  Microfono Para Pc Nisuta Ns-mic130 Con Base 50...    837.00   

       sale_price currency_id  available_quantity  sold_quantity buying_mode  \
0             NaN         ARS                  83           3179  buy_it_now   
1             NaN         ARS                  40           5858  buy_it_now   
2             NaN         ARS                 250            500  buy_it_now   
3             NaN         ARS                 500           5000  buy_it_now   
4             NaN         ARS                 500            500  buy_it_now   
...           ...         ...                 ...            ...         ...   
10478         NaN         ARS                   1            500  buy_it_now   
10479         NaN         ARS                   1            100  buy_it_now   
10480         NaN         ARS                 500            100  buy_it_now   
10481         NaN         ARS                   1            100  buy_it_now   
10482         NaN         ARS                   1             25  buy_it_now   

      listing_type_id                 stop_time condition  \
0        gold_special  2040-12-07T04:00:00.000Z       new   
1        gold_special  2040-05-13T04:00:00.000Z       new   
2            gold_pro  2037-08-24T02:40:31.000Z       new   
3            gold_pro  2040-05-07T04:00:00.000Z       new   
4        gold_special  2040-02-23T04:00:00.000Z       new   
...               ...                       ...       ...   
10478    gold_special  2037-08-09T18:28:24.000Z       new   
10479        gold_pro  2040-08-14T19:20:06.000Z       new   
10480    gold_special  2040-06-04T04:00:00.000Z       new   
10481    gold_special  2039-10-03T04:00:00.000Z       new   
10482    gold_special  2040-05-16T04:00:00.000Z       new   

                                               permalink  \
0      https://www.mercadolibre.com.ar/repetidor-tp-l...   
1      https://www.mercadolibre.com.ar/disco-solido-i...   
2      https://articulo.mercadolibre.com.ar/MLA-61888...   
3      https://articulo.mercadolibre.com.ar/MLA-85505...   
4      https://articulo.mercadolibre.com.ar/MLA-84142...   
...                                                  ...   
10478  https://articulo.mercadolibre.com.ar/MLA-67867...   
10479  https://articulo.mercadolibre.com.ar/MLA-85882...   
10480  https://articulo.mercadolibre.com.ar/MLA-86112...   
10481  https://articulo.mercadolibre.com.ar/MLA-81875...   
10482  https://articulo.mercadolibre.com.ar/MLA-85701...   

                                               thumbnail  \
0      http://http2.mlstatic.com/D_668117-MLA31786420...   
1      http://http2.mlstatic.com/D_971286-MLA40103563...   
2      http://http2.mlstatic.com/D_856328-MLA42504174...   
3      http://http2.mlstatic.com/D_621392-MLA43401053...   
4      http://http2.mlstatic.com/D_975747-MLA40988335...   
...                         

### Analisis preliminar. 
##### Se eliminan todas las columnas que no aporten información. 
##### Ver análisis preliminar en MLC_04_ANALISIS_PRELIMINAR.ipynb

### Limpieza preliminar del Dataset
#### Teniendo en cuenta el resultado del análisis preliminar se decide eliminar los campos que no aporten información. 
##### (Campos que sean todos NULL, o que contengan todos el mismo valor. )

In [20]:
to_drop_list = ["site_id", "sale_price", "currency_id", "buying_mode", "condition", "permalink",
               "thumbnail_id", "thumbnail", "accepts_mercadopago", "catalog_listing", "order_backend",
               "seller.permalink", "seller.car_dealer", "seller.real_estate_agency",
               "seller.eshop.eshop_rubro", "seller.eshop.eshop_id", "seller.eshop.eshop_locations", 
               "seller.eshop.site_id", "seller.eshop.eshop_logo_url", "seller.eshop.seller", "seller.eshop.eshop_experience",
               "seller.seller_reputation.transactions.period", "seller.seller_reputation.metrics.claims.period", "seller.seller_reputation.metrics.delayed_handling_time.period",
               "seller.seller_reputation.metrics.delayed_handling_time.period", "seller.seller_reputation.metrics.cancellations.period", "seller.seller_reputation.level_id", 
               "prices.presentation.display_currency", "prices.payment_method_prices", "installments.currency_id", "address.state_id", "address.city_id", 
               "shipping.mode", "shipping.store_pick_up", "seller_address.id", "seller_address.address_line", "seller_address.comment",
               "seller_address.zip_code", "seller_address.country.id", "seller_address.country.name", "seller_address.state.id",
               "seller_address.state.name", "seller_address.city.id", "seller_address.city.name", "seller_address.latitude",
               "seller_address.longitude", "seller_address.longitude", "seller.seller_reputation.metrics.sales.period",
               "attributes", "prices.prices", "shipping.tags"]

# La mayoría de las columnas se agregaron a esta lista por no tener información relevante. 
# Algunas se agregaron por tener información repetida (seller.address.*). 
# "attributes", "prices.prices", "shipping.tags" se agregaron arbitrariamente.

In [21]:
raw_dataset_df.drop(axis=1, labels=to_drop_list, inplace=True)

In [22]:
raw_dataset_df

id                                              title  \
0      MLA900810750      Repetidor Tp-link Tl-wa850re  Blanco 1 Unidad   
1      MLA856313404  Disco Sólido Interno Kingston Sa400s37/240g 240gb   
2      MLA618887655  Pc Armada Cpu Amd Full Gamer A10-9700 16gb Ddr...   
3      MLA855055612  Notebook Exo Smart E19 Intel Celeron Ram4gb Ss...   
4      MLA841426834  Pasta Térmica Arctic Mx-4 4grs. Alto Rendimien...   
...             ...                                                ...   
10478  MLA678679714  Auriculares Genius Hs 05a Headset Control Volu...   
10479  MLA858821035  The Lego Movie Videogame Formato Físico Ps4 Or...   
10480  MLA861128302  Call Of Duty Black Ops 2 Ps3 Gold | Español | ...   
10481  MLA818759510  Cable Cargador Y Transferencia Ps Vita Alto Re...   
10482  MLA857013842  Microfono Para Pc Nisuta Ns-mic130 Con Base 50...   

          price  available_quantity  sold_quantity listing_type_id  \
0       2055.00                  83           3179    gold_special   
1       3688.00                  40           5858    gold_special   
2      53999.00                 250            500        gold_pro   
3      39999.00                 500           5000        gold_pro   
4       1600.00                 500            500    gold_special   
...         ...                 ...            ...             ...   
10478   2249.00                   1            500    gold_special   
10479   2999.00                   1            100        gold_pro   
10480    449.95                 500            100    gold_special   
10481    388.99                   1            100    gold_special   
10482    837.00                   1             25    gold_special   

                      stop_time  original_price category_id  \
0      2040-12-07T04:00:00.000Z             NaN   MLA430901   
1      2040-05-13T04:00:00.000Z             NaN     MLA1672   
2      2037-08-24T02:40:31.000Z         74999.0     MLA1649   
3      2040-05-07T04:00:00.000Z         46799.0     MLA1652   
4      2040-02-23T04:00:00.000Z             NaN     MLA2676   
...                         ...             ...         ...   
10478  2037-08-09T18:28:24.000Z             NaN     MLA6049   
10479  2040-08-14T19:20:06.000Z             NaN   MLA373840   
10480  2040-06-04T04:00:00.000Z             NaN   MLA373840   
10481  2039-10-03T04:00:00.000Z             NaN   MLA121969   
10482  2040-05-16T04:00:00.000Z             NaN     MLA1664   

       official_store_id                              domain_id  \
0                    NaN       MLA-ROUTERS_AND_WIRELESS_SYSTEMS   
1                    NaN               MLA-HARD_DRIVES_AND_SSDS   
2                    NaN                  MLA-DESKTOP_COMPUTERS   
3                  341.0                          MLA-NOTEBOOKS   
4                    NaN  MLA-DESKTOP_COMPUTER_COOLERS_AND_FANS   
...                  ...                                    ...   
10478             1178.0                         MLA-HEADPHONES   
10479                NaN                        MLA-VIDEO_GAMES   
10480                NaN                        MLA-VIDEO_GAMES   
10481                NaN     MLA-VIDEO_GAME_CONTROLLER_CHARGERS   
10482                NaN                        MLA-MICROPHONES   

      catalog_product_id                                               tags  \
0             MLA6186903  ['brand_verified', 'good_quality_picture', 'im...   
1             MLA6408981  ['brand_verified', 'good_quality_picture', 'go...   
2                    NaN  ['ahora-12', 'brand_verified', 'extended_warra...   
3            MLA16368693  ['good_quality_thumbnail', 'ahora-12', 'brand_...   
4                    NaN  ['brand_verified', 'good_quality_picture', 'go...   
...                  ...                                                ...   
10478        MLA11339074  ['brand_verified', 'good_quality_picture', 'go...   
10479         MLA6186755  ['brand_verified', 'good_quality_picture', 'go...   
10480        M

##### Se redujo el dataset de 102 columnas, a 53 (que pueden o no tener información útil)

# Análisis específico de Items co